src/Fx.tsx

```diff
@@ -0,0 +1,245 @@
+        for (let i = 0; i < 600; i++) {
+            const angleDeviation = Math.PI * 0.8  * (Math.random() - 0.5);
+            const powerDeviation = -0.95 * Math.random() * power;
+            this.particleSystem.addParticle(x, y, Math.random() * Math.PI * 2, Math.cos(angle + angleDeviation) * (power + powerDeviation), Math.sin(angle + angleDeviation) * (power + powerDeviation), (Math.random() - 0.5) * 4);
+        }
+    }
+
+    public render(gfx: CanvasRenderingContext2D) {
+        for (const { idx, ttl, x, y, angle } of this.particleSystem.listParticles()) {
+            const alpha = Math.min(ttl, 1.0);
+            gfx.fillStyle = `rgba(${[
+                '168,100,253',
+                '41,205,255',
+                '120,255,68',
+                '255,113,141',
+                '253,255,106',
+            ][idx % 5]},${alpha})`;
+            const xOffset = Math.sin(angle) * 30;
+            gfx.translate(x + xOffset, y);
+            gfx.rotate(angle);
+            gfx.fillRect(-5, -3, 10, 6);
+            gfx.resetTransform();
+        }
+    }
+}
+
+type FxState = {
+    canvas: HTMLCanvasElement,
+    gfx: CanvasRenderingContext2D,
+    emitters: Array<Emitter>,
+    active: boolean,
+};
+
+type FxContext = {
+    shootConfetti: (x: number, y: number, angle: number) => void;
+};
+
+export const FxContext = React.createContext<FxContext>({shootConfetti: () => {}});
+
+export const FxProvider: React.FC<React.PropsWithChildren> = ({children}) => {
+    const canvasRef = useRef<HTMLCanvasElement>(null);
+    const confettiRef = useRef<(x: number, y: number, angle: number) => void>();
+    useEffect(() => {
+        if (canvasRef.current === null) {
+            return;
+        }
+        const fx = createFxContext(canvasRef.current);
+        if (fx === null) {
+            return;
+        }
+        const confettiEmitter = new ConfettiEmitter();
+        fx.emitters.push(confettiEmitter);
+        confettiRef.current = (x: number, y: number, angle: number) => {
+            confettiEmitter.shootConfetti(x, y, angle, 18_000);
+        };
+        startFxLoop(fx);
+        return () => {
+            fx.active = false;
+        }
+    }, []);
+
+    const shootConfetti = useCallback((x: number, y: number, angle: number) => {
+        confettiRef.current?.(x, y, angle)
+    }, [confettiRef])
+
+    return (
+        <FxContext.Provider value={{ shootConfetti }}>
+            <canvas ref={canvasRef} style={{top: 0, left: 0, position: "fixed", pointerEvents: "none"}}></canvas>
+            {children}
+        </FxContext.Provider>
+    );
+};
+
+function startFxLoop(fx: FxState) {
+    let lastRenderMs = Date.now();
+    const loopImpl = () => {
+        if (!fx.active) {
+            return;
+        }
+        const nowMs = Date.now();
+        const timeDeltaMs = nowMs - lastRenderMs;
+        lastRenderMs = nowMs;
+        renderFxFrame(fx, timeDeltaMs / 1_000);
+        setTimeout(loopImpl, 1_000 / FX_FPS)
```

Comment:
This should be using requestAnimationFrame

Reply:
hmmm, does that limit framerate to 60, or based on display refresh rate? I was hoping to always just limit to 60

Reply:
Why fps cap

Reply:
it's a silly effect; don't want it comsuming too much resources

Reply:
requestAnimationFrame also handles other edge cases like weird (or lower) refresh rates, doesn't run while offscreen, and provides a more accurate timestamp for smooth animation though.
You could maybe skip the frame if it's too close to the last one

Reply:
ah good idea